src/client/datascience/crossProcessLock.ts

```
@@ -0,0 +1,58 @@
+import { open, unlink } from 'fs-extra';
+import { tmpdir } from 'os';
+import * as path from 'path';
+import { sleep } from '../common/utils/async';
+
+export class CrossProcessLock {
+    private lockFilePath: string;
+    private acquired: boolean = false;
+
+    constructor(mutexName: string) {
+        this.lockFilePath = path.join(tmpdir(), `${mutexName}.tmp`);
+    }
+
+    public async lock(): Promise<boolean> {
+        const maxTries = 10;
+        let tries = 0;
+        while (!this.acquired && tries < maxTries) {
+            try {
+                await this.acquire();
+                if (this.acquired) {
+                    return true;
+                }
+                await sleep(100);
+            } catch (err) {
+                // Swallow the error and retry
+            }
+            tries += 1;
+        }
+        return false;
+    }
+
+    public async unlock() {
+        // Does nothing if the lock is not currently held
+        if (this.acquired) {
```

I'd throw an error if we call `unlock` when we haven't acquired a lock.